rfcs/0007-parametrized-unit-tests.md

```diff
@@ -0,0 +1,568 @@
+cache_setup1 = {}
+cache_setup2 = {}
+for array_size in array_sizes:
+    for target in targets:
+        if array_size in cache_setup1:
+            setup1 = cache_setup1[array_size]
+        else:
+            setup1 = cache_setup1[array] = generate_setup1(array_size)
+
+        if target in cache_setup2:
+            setup2 = cache_setup2[target]
+        else:
+            setup2 = cache_setup2[target] = generate_setup2(target)
+
+        run_test(setup1, setup2)
+
+del cache_setup1
+del cache_setup2
+```
+
+The cache for a fixture defined with `tvm.testing.fixture` is cleared
+after all tests using that fixture are completed, to avoid excessive
+memory usage.
+
+If a test function is marked with `@pytest.mark.parametrize` for a
+parameter that is also defined with `tvm.testing.parameter`, the test
+function uses only the parameters in `@pytest.mark.parametrize`.  This
+allows an individual function to override the parameter definitions if
+needed.  Any parameter-dependent fixture are also determined based on
+the values in `@pytest.mark.parametrize`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This makes the individual unit tests be more dependent on the test
+  framework and setup.  Incorrect setup may result in confusing test
+  results.
+
+- Caching setup between different tests introduces potential
+  cross-talk between tests.  While this risk is also present when
+  looping over parameter values, separating cached values out into
+  fixtures hides that potential cross-talk.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Option: Explicitly loop over parameter values or
+  `tvm.testing.enabled_parameters` in the test function.  (Most common
+  previous usage.)
+  
+  - Pros:
+    - Explicit at the definition of a test function.
+    
+  - Cons:
+    - Requires opt-in at each test functions.
+    - Doesn't report information on which parameter value(s) failed.
+    
+    
+- Option: Use `@tvm.testing.parametrize_targets` as a bare fixture.
+  (Previously implemented behavior, less common usage.)
+
+  - Pros:
+    - Explicit at the definition of a test function.
+    
+  - Cons:
+    - Requires opt-in at each test function.
+    - Doesn't provide functionality for shared setup.
+    
+
+- Option: Pararametrize using `@pytest.mark.parametrize` rather than
+  `tvm.testing.parameter`.
+  
+  - Pros:
+    - Would explicitly show the parameter values next to the function
+      it applies to.
+      
+  - Cons:
+    - Must be explicitly added at each test function definition.
+    - Extending the parameters that apply across all tests in a
+      file/directory requires updating several locations.
+    - Similar parameters (e.g. 1000 vs 1024 for an array length) would
+      be defined at separate locations, and would then require
+      separate fixture setup.
+
+# Prior art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
```

Comment:
can you remove the boilerplate and just add a reference to i guess `pytest.mark.parametrize` and `functools.lru_cache` or whatever that is now?

Reply:
Updated with a reference to `pytest.mark.parametrize` and `pytest.fixture`'s caching using the `scope` parameter.